crates/astria-sequencer/src/action_handler/impls/recover_ibc_client.rs

```diff
@@ -0,0 +1,167 @@
+            .wrap_err("failed to get timestamp")?;
+        let client_status = state.get_client_status(&self.client_id, timestamp).await;
+
+        ensure!(
+            client_status != ClientStatus::Active,
+            "cannot recover an active client",
+        );
+
+        let replacement_client_status = state
+            .get_client_status(&self.replacement_client_id, timestamp)
+            .await;
+
+        ensure!(
+            replacement_client_status == ClientStatus::Active,
+            "substitute client must be active: status is {}",
+            replacement_client_status,
+        );
+
+        let mut client_state = state
+            .get_client_state(&self.client_id)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("subject client state not found")?;
+        let replacement_client_state = state
+            .get_client_state(&self.replacement_client_id)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("substitute client state not found")?;
+
+        ensure!(
+            client_state.latest_height() < replacement_client_state.latest_height(),
+            "substitute client must have a higher height than that of subject client; subject \
+             client height: {}, substitute client height: {}",
+            client_state.latest_height(),
+            replacement_client_state.latest_height(),
+        );
+
+        ensure_required_client_state_fields_match(&client_state, &replacement_client_state)?;
+
+        let height = ibc_types::core::client::Height {
+            revision_height: state
+                .get_block_height()
+                .await
+                .wrap_err("failed to get block height")?,
+            revision_number: state
+                .get_revision_number()
+                .await
+                .wrap_err("failed to get revision number")?,
+        };
+        let substitute_consensus_state = state
+            .prev_verified_consensus_state(&self.replacement_client_id, &height)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("failed to get substitute client consensus state")?
+            .ok_or_eyre("substitute client consensus state not found")?;
+        state
+            .put_verified_consensus_state::<crate::ibc::host_interface::AstriaHost>(
+                replacement_client_state.latest_height(),
+                self.client_id.clone(),
+                substitute_consensus_state,
+            )
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("failed to put verified consensus state")?;
+
+        client_state.latest_height = replacement_client_state.latest_height;
+        client_state.trusting_period = replacement_client_state.trusting_period;
+        client_state.chain_id = replacement_client_state.chain_id;
+        state.put_client(&self.client_id, client_state);
+        Ok(())
+    }
+}
+
+fn ensure_required_client_state_fields_match(
```

Comment:
nit: probably add a comment above which notes the spec which defines what checks are needed?